src/transformers/models/t5/modeling_t5.py

```diff
@@ -919,71 +1059,160 @@ def forward(
+        present_key_values_p1 = torch.empty(batch_size, n_layers * 2).to(input_ids.device)
+        present_key_values_p2 = torch.empty(batch_size, n_layers * 2).to(input_ids.device)
+
+        # rewrite the model after pre-trained weights were loaded
+        layers = [
+            T5StackPipeSegment(
+                idx,
+                n_layers,
+                layer_module,
+                self.is_decoder,
+                head_mask[idx],
+                encoder_head_mask[idx],
+                output_hidden_states,
+                use_cache,
+                output_attentions,
+                all_hidden_states_add,
+                present_key_value_states_add,
+                all_attentions_add,
+                all_cross_attentions_add,
             )
-            # layer_outputs is a tuple with:
-            # hidden-states, key-value-states, (self-attention weights), (self-attention position bias), (cross-attention weights), (cross-attention position bias)
-            hidden_states, present_key_value_state = layer_outputs[:2]
-
-            # We share the position biases between the layers - the first layer store them
-            # layer_outputs = hidden-states, key-value-states (self-attention weights),
-            # (self-attention position bias), (cross-attention weights), (cross-attention position bias)
-            position_bias = layer_outputs[2]
-            if self.is_decoder and encoder_hidden_states is not None:
-                encoder_decoder_position_bias = layer_outputs[4 if output_attentions else 3]
-            # append next layer key value states
-            if use_cache:
-                present_key_value_states = present_key_value_states + (present_key_value_state,)
-
-            if output_attentions:
-                all_attentions = all_attentions + (layer_outputs[3],)
-                if self.is_decoder:
-                    all_cross_attentions = all_cross_attentions + (layer_outputs[5],)
-
-            # Model Parallel: If it's the last layer for that device, put things on the next device
-            if self.model_parallel:
-                for k, v in self.device_map.items():
-                    if i == v[-1] and "cuda:" + str(k) != self.last_device:
-                        hidden_states = hidden_states.to("cuda:" + str(k + 1))
+            for idx, layer_module in enumerate(self.block)
+        ]
+        # block_sequential = nn.Sequential(*layers)
+
+        # for now don't enable the pipe
+        if self.pipeline_is_enabled:
+
+            # print("using partitioning: ", dict(zip(devices, layer_splits)))
+            for device_id, layer_partition in self.device_map.items():
+                for layer_id in layer_partition:
+                    # print(f"{layer_id} => {device_id}")
+                    layers[layer_id].to(device_id)
+
+            block_sequential = nn.Sequential(*layers)
+            block_pipe = Pipe(block_sequential, chunks=self.pipeline_chunks, checkpoint="never")
```

Comment:
because if you look a bit higher where I create `T5StackPipeSegment` to populate the layers I don't have all those arguments available at init time - i.e. I only get those during forward. But I can pass them to forward of course. I was just trying to come up with ways to overcome the Tensor/tuple(Tensor) restriction and at the time  this seems like a good workaround. If I don't do it, the passing of arguments to forward becomes even more complicated.

But as you started looking into in order to identify why there is no speed up over naive MP, it could be the cause - I have no idea how long does it take to build these 2 pipes - but surely it's expensive doing it on every model run.

My first experiment was just to make the pipe work at all costs, so clearly this is not a way forward.

Reply:
And btw, continuing our discussion from the pytorch issue, we also can't easily change the shape of the model, because we need to adhere to the pre-trained model layout to be able to load pre-trained models. I suppose it could be re-mapped after the normal dist state is loaded. So that's another complication I was trying to overcome here - injecting proxy layers that had no weights of its own.